README.md
# Capstan

![GitHub Rust CI](https://github.com/lancelet/capstan/workflows/Rust/badge.svg)
[![Codecov.io](https://codecov.io/gh/lancelet/capstan/branch/main/graph/badge.svg)](https://codecov.io/gh/lancelet/capstan)
[![Crates.io](https://img.shields.io/crates/v/capstan.svg)](https://crates.io/crates/capstan)
[![Docs.rs](https://docs.rs/capstan/badge.svg)](https://docs.rs/capstan)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

NURBS utilities in Rust.

## NURBS Curve Evaluation

Currently, only NURBS curve evaluation is complete. The evaluation uses a
naive version of the de Boor algorithm. With this, it's possible to evaluate
the 3D coordinates of a NURBS curve at any parameter value.

NURBS can represent conics with floating-point precision. This image shows a
tesselated NURBS circle on the left and an SVG circle on the right:

<img src="./diagrams/circle.svg" width="600" height="300"/>

NURBS are a generalization of Bézier curves, so they can exactly represent any
order of Bézier curve. The image below shows an SVG cubic Bézier with a loop on
the right and a tesselated NURBS representation on the left:

<img src="./diagrams/cubic-bezier.svg" width="600" height="300"/>

NURBS can represent multiple Bézier curve segments in a single curve. The
example below shows an outline of the Egyptian "reed leaf" hieroglyph
(Gardiner sign M17). This curve is constructed from 2 line segments and 4
cubic Bézier curves, all of which can be represented as a single closed
NURBS curve:

<img src="./diagrams/reed-leaf.svg" width="300" height="300"/>

## NURBS Curve Representation

The library uses the "Rhino" form of NURBS curves, where there are two fewer
knots than in "traditional" NURBS.

## Usage

A simple circle example demonstrating how to obtain an interpolated value at an arbitrary location on a curve.

```rust
use nalgebra::Vector2;
use capstan::{knotvec::KnotVec};
type Curve = capstan::curve::Curve<f32, Vector2<f32>>;

fn main() {

    let r = f32::sqrt(2.0) / 2.0;
    let degree = 2;
    let control_points = vec![
        Vector2::new(1.0, 0.0),
        Vector2::new(1.0, 1.0),
        Vector2::new(0.0, 1.0),
        Vector2::new(-1.0, 1.0),
        Vector2::new(-1.0, 0.0),
        Vector2::new(-1.0, -1.0),
        Vector2::new(0.0, -1.0),
        Vector2::new(1.0, -1.0),
        Vector2::new(1.0, 0.0),
    ];
    let weights = vec![1.0, r, 1.0, r, 1.0, r, 1.0, r, 1.0];
    let knots = KnotVec::new(vec![
        0.0, 0.0, 0.0, 0.25, 0.25, 0.5, 0.5, 0.75, 0.75, 1.0, 1.0, 1.0,
    ])
    .unwrap();

    let circle = Curve::new(degree, control_points, weights, knots).unwrap();

    let u = 0.5_f32;

    // The de_boor method is used to obtainin an interpolated point
    println!("interpolation paramater value {}:{}", u, circle.de_boor(u))
}
```

## Examples

The examples folder includes the above as well as the more comprehensive svg_example.rs example that outputs a number of svg files. The svg crate is used for this but is only required for demonstration purposes.

To run the svg_example:
$ cargo run --example svg_example

This will output a number of .svg files into the working directory

CHANGELOG.md
# Changelog

Please track all notable changes in this file. This format is based on
[Keep a Changelog](https://keepachangelog.com/en/1.0.0).

## [Unreleased]

## [0.0.4]

### Added

- Added short circle example directly to the README as well as to a new examples folder

### Changed

- Bumped rust edition to 2021
- Upgraded the Nalbegra dependency to 0.29
- Implemented VectorT for OVector instead of VectorN as VectorN is now deprecated.
- Moved main.rs to the examples folder
- Moved the svg and approx to dev-dependencies as these are used in examples / testing only.
- Edited .gitignore to ignore root level .svg artifacts generated from the example

## [0.0.3]

### Added

- `KnotVec.is_clamped` function.
- "Reed leaf hieroglyph" example of a curve with multiple Bézier segments.

### Changed

- Changed the representation to use full-multiplicity knots instead of the
  "Rhino" style.
- `KnotVec` is now passed to `Curve` on creation, instead of an arbitrary
  `Vec`.

## [0.0.2]

### Added

- `ScalarT` and `VectorT` traits to represent the operations required of
  scalars and vectors used by NURBS curves and surfaces.
- `KnotVec` type to represent knot vectors in NURBS curves. This will later be
  shared with NURBS surfaces.
- `codecov.io` test coverage using `grcov`.
- README badges for `codecov.io`, `crates.io`, `docs.rs` and the LICENSE.

### Changed

- `Curve` type is now parameterised by scalar and vector types.
- Examples in `main` use `nalgebra::Vector2` instead of 3D points.

## [0.0.1]

### Added

- Initial project setup.
- NURBS Curve representation and evaluation using the de Boors algorithm.
- Plot some examples for the README.

[unreleased]: https://github.com/lancelet/capstan/compare/v0.0.3...HEAD
[0.0.3]: https://github.com/lancelet/capstan/releases/tag/v0.0.3
[0.0.2]: https://github.com/lancelet/capstan/releases/tag/v0.0.2
[0.0.1]: https://github.com/lancelet/capstan/releases/tag/v0.0.1

src/curve.rs
use crate::algebra::{ScalarT, VectorT};
use crate::knotvec::KnotVec;
use thiserror::Error;

pub type Result<T> = std::result::Result<T, CurveError>;

/// NURBS curve.
///
/// Non-Uniform Rational B-Spline.
#[derive(PartialEq, Debug)]
pub struct Curve<N, V>
where
    N: ScalarT,
    V: VectorT<Field = N>,
{
    degree: usize,
    control_points: Vec<V>,
    weights: Vec<N>,
    knots: KnotVec<N>,
}

impl<N, V> Curve<N, V>
where
    N: ScalarT,
    V: VectorT<Field = N>,
{
    /// Creates a new NURBS Curve.
    ///
    /// The following basic properties must be satisfied for a NURBS curve:
    /// * `degree` > 0
    /// * `control_points.len() > degree`
    /// * `weights.len() == control_points.len()`
    /// * `knots.len() == degree + control_points.len() + 1`
    /// * `knots.is_clamped()`
    ///
    /// The NURBS curves represented here are clamped (ie. they must have a
    /// knot multiplicity at either end equal to the degree plus one).
    /// Un-clamped curves can be converted to clamped ones via knot insertion.
    ///
    /// Parameters:
    ///
    /// * `degree` - polynomial degree of the NURBS curve
    /// * `control_points` - vector of control points
    /// * `weights` - vector of weights (must be the same length as
    ///               `control_points`)
    /// * `knots` - knot vector (must have `degree + control_points.len() + 1`
    ///             elements)
    pub fn new(
        degree: usize,
        control_points: Vec<V>,
        weights: Vec<N>,
        knots: KnotVec<N>,
    ) -> Result<Self> {
        if degree == 0 {
            Err(CurveError::InvalidDegree)
        } else if control_points.len() <= degree {
            Err(CurveError::InsufficientControlPoints {
                degree,
                number_supplied: control_points.len(),
            })
        } else if weights.len() != control_points.len() {
            Err(CurveError::MismatchedWeightsAndControlPoints)
        } else if knots.len() != degree + control_points.len() + 1 {
            Err(CurveError::InvalidKnotCount {
                required_knot_len: degree + control_points.len() + 1,
                receieved_knot_len: knots.len(),
            })
        } else if !knots.is_clamped(degree) {
            Err(CurveError::KnotVectorNotClamped)
        } else {
            Ok(Curve {
                degree,
                control_points,
                weights,
                knots,
            })
        }
    }

    /// Interpolates the curve at a parameter value.
    ///
    /// This method uses the
    /// [de Boor algorithm](https://en.wikipedia.org/wiki/De_Boor%27s_algorithm)
    /// to evaluate the NURBS curve at a given parameter value `u`. The de Boor
    /// algorithm is a good choice for efficiently evaluating a NURBS curve and
    /// is numerically stable.
    ///
    /// The parameter `u` is clamped to the allowed range of the parameter
    /// space of the curve (which is the range from `self.knots().min_u()` to
    /// `self.knots().max_u()` inclusive).
    ///
    /// # Parameters
    ///
    /// * `u` - the parameter value at which to evaluate the NURBS curve
    pub fn de_boor(&self, u: N) -> V {
        // clamp u and find the knot span containing u
        let uu = self.knots.clamp(u);
        let k = self.knots.find_span(uu);

        // populate initial triangular column
        let mut d = Vec::<V>::with_capacity(self.degree + 1); // homogeneous points
        let mut dw = Vec::<N>::with_capacity(self.degree + 1); // weights
        for j in 0..self.degree + 1 {
            let i: usize = j + k - self.degree;

            // multiply the control points by the corresponding weight to
            // convert from Cartesian to homogeneous coordinates
            d.push(self.control_points[i].clone() * self.weights[i]);
            dw.push(self.weights[i]);
        }

        // make extra-sure we allocated enough capacity
        debug_assert!(d.len() <= self.degree + 1);
        debug_assert!(dw.len() <= self.degree + 1);

        // main de Boor algorithm
        for r in 1..self.degree + 1 {
            for j in (r..self.degree + 1).rev() {
                let kp = self.knots[j + k - self.degree];
                let alpha = (uu - kp) / (self.knots[1 + j + k - r] - kp);
                let nalpha = N::one() - alpha;
                d[j] = d[j - 1].clone() * nalpha + d[j].clone() * alpha;
                dw[j] = dw[j - 1] * nalpha + dw[j] * alpha;
            }
        }

        // convert final coordinate from homogeneous to Cartesian coords
        d[self.degree].clone() * (N::one() / dw[self.degree])
    }

    /// Returns the vector of control points.
    pub fn control_points(&self) -> &Vec<V> {
        &self.control_points
    }

    /// Returns the knot vector.
    pub fn knots(&self) -> &KnotVec<N> {
        &self.knots
    }

    /// Scale the curve by a uniform amount about the origin.
    ///
    /// NOTE: This method will probably be replaced by a more general
    ///       transformation method in the future.
    pub fn uniform_scale(&mut self, scale_factor: N) {
        for cp in &mut self.control_points {
            *cp = cp.clone() * scale_factor;
        }
    }
}

#[derive(Error, Debug, PartialEq)]
pub enum CurveError {
    #[error("invalid degree; must satisfy degree > 0")]
    InvalidDegree,

    #[error("N={} control points were supplied; at least {} are required \
             for a degree {} curve",
            .number_supplied,
            .degree,
            .degree - 1)]
    InsufficientControlPoints {
        degree: usize,
        number_supplied: usize,
    },

    #[error("number of weights and control points must be identical")]
    MismatchedWeightsAndControlPoints,

    #[error("expected {} knot values, but received {}",
            .required_knot_len,
            .receieved_knot_len)]
    InvalidKnotCount {
        required_knot_len: usize,
        receieved_knot_len: usize,
    },

    #[error("knot vector was not clamped")]
    KnotVectorNotClamped,
}

#[cfg(test)]
mod tests {
    use super::*;
    use approx::assert_relative_eq;
    use nalgebra::Vector2;

    /// Test Curve
    type TC = Curve<f32, Vector2<f32>>;

    /// The degree of a NURBS curve must be >= 0.
    #[test]
    fn invalid_degree() {
        let result = TC::new(0, vec![], vec![], KnotVec::new(vec![0.0, 1.0]).unwrap());
        assert_eq!(result, Err(CurveError::InvalidDegree));
    }

    /// There must be at least degree + 1 control points.
    #[test]
    fn insufficient_control_points() {
        let result = TC::new(
            1,
            vec![Vector2::new(0.0, 0.0)],
            vec![1.0],
            KnotVec::new(vec![0.0, 0.0, 1.0, 1.0]).unwrap(),
        );
        assert_eq!(
            result,
            Err(CurveError::InsufficientControlPoints {
                degree: 1,
                number_supplied: 1
            })
        )
    }

    /// The number of control points and weights must be identical.
    #[test]
    fn weights_and_cps_lengths_must_be_equal() {
        let result = TC::new(
            1,
            vec![Vector2::new(0.0, 0.0), Vector2::new(42.0, 56.0)],
            vec![1.0],
            KnotVec::new(vec![0.0, 1.0]).unwrap(),
        );
        assert_eq!(result, Err(CurveError::MismatchedWeightsAndControlPoints));
    }

    /// The correct number of knot values must be supplied.
    #[test]
    fn invalid_knot_count() {
        let result = TC::new(
            1,
            vec![Vector2::new(0.0, 0.0), Vector2::new(42.0, 56.0)],
            vec![1.0, 1.0],
            KnotVec::new(vec![0.0, 1.0]).unwrap(),
        );
        assert_eq!(
            result,
            Err(CurveError::InvalidKnotCount {
                required_knot_len: 4,
                receieved_knot_len: 2
            })
        );
    }

    /// Test that we detect a non-clamped knot vector.
    #[test]
    fn knot_vector_not_clamped() {
        let result = TC::new(
            2,
            vec![
                Vector2::new(0.0, 0.0),
                Vector2::new(1.0, 2.0),
                Vector2::new(3.0, 4.0),
            ],
            vec![1.0, 1.0, 1.0],
            KnotVec::new(vec![0.0, 0.0, 0.5, 0.5, 0.9, 1.0]).unwrap(),
        );
        assert_eq!(result, Err(CurveError::KnotVectorNotClamped));
    }

    /// Creating a new NURBS curve successfully.
    #[test]
    fn new() {
        let nurbs = TC::new(
            1,
            vec![Vector2::new(0.0, 0.0), Vector2::new(42.0, 56.0)],
            vec![1.0, 1.0],
            KnotVec::new(vec![0.0, 0.0, 1.0, 1.0]).unwrap(),
        )
        .unwrap();
        assert_eq!(nurbs.knots().min_u(), 0.0);
        assert_eq!(nurbs.knots().max_u(), 1.0);
        assert_eq!(
            nurbs.control_points(),
            &vec![Vector2::new(0.0, 0.0), Vector2::new(42.0, 56.0)]
        );
    }

    /// Uniformly scaling a NURBS curve.
    #[test]
    fn uniform_scale() {
        let mut nurbs = TC::new(
            1,
            vec![Vector2::new(0.0, 0.0), Vector2::new(42.0, 56.0)],
            vec![1.0, 1.0],
            KnotVec::new(vec![0.0, 0.0, 1.0, 1.0]).unwrap(),
        )
        .unwrap();
        nurbs.uniform_scale(2.0);

        let expected = TC::new(
            1,
            vec![Vector2::new(0.0, 0.0), Vector2::new(2.0 * 42.0, 2.0 * 56.0)],
            vec![1.0, 1.0],
            KnotVec::new(vec![0.0, 0.0, 1.0, 1.0]).unwrap(),
        )
        .unwrap();

        assert_eq!(nurbs, expected);
    }

    /// Test de Boor evalutaion on a non-rational, uniform Bezier.
    #[test]
    fn de_boor_non_rational_uniform_bezier() {
        let test_curve = TC::new(
            3,
            vec![
                Vector2::new(-10.0, 10.0),
                Vector2::new(10.0, 10.0),
                Vector2::new(-10.0, -10.0),
                Vector2::new(10.0, -10.0),
            ],
            vec![1.0, 1.0, 1.0, 1.0],
            KnotVec::new(vec![0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0]).unwrap(),
        )
        .unwrap();

        // tests for in-range parameter
        assert_relative_eq!(Vector2::new(-10.0, 10.0), test_curve.de_boor(0.0));
        assert_relative_eq!(Vector2::new(-2.16, 7.92), test_curve.de_boor(0.2));
        assert_relative_eq!(Vector2::new(0.0, 0.0), test_curve.de_boor(0.5));
        assert_relative_eq!(Vector2::new(10.0, -10.0), test_curve.de_boor(1.0));

        // tests with parameter out-of-range (clipped to parameter range)
        assert_relative_eq!(Vector2::new(-10.0, 10.0), test_curve.de_boor(-1.0));
        assert_relative_eq!(Vector2::new(10.0, -10.0), test_curve.de_boor(2.0));
    }
}


src/algebra.rs
use nalgebra::OVector;
use nalgebra::base::allocator::Allocator;
use nalgebra::base::{DefaultAllocator, DimName};
use num_traits::identities::One;
use std::fmt::Debug;
use std::ops::{Add, AddAssign, Div, Mul, MulAssign, Sub};

/// A scalar type.
///
/// Scalars are used for things like knot locations, weights, parameter values,
/// and the scalar components of vector types.
pub trait ScalarT:
    Copy
    + PartialOrd
    + Debug
    + Add<Output = Self>
    + AddAssign
    + Mul<Output = Self>
    + MulAssign
    + Sub<Output = Self>
    + Div<Output = Self>
    + One
{
}

impl<T> ScalarT for T where
    T: Copy
        + PartialOrd
        + Debug
        + Add<Output = Self>
        + AddAssign
        + Mul<Output = Self>
        + MulAssign
        + Sub<Output = Self>
        + Div<Output = Self>
        + One
{
}

/// A vector type.
///
/// Vectors are used for 3D locations like control points and points on curves
/// or surfaces.
pub trait VectorT:
    Clone + Debug + Add<Output = Self> + Mul<<Self as VectorT>::Field, Output = Self>
{
    type Field: ScalarT;
}

impl<N, D> VectorT for OVector<N, D>
where
    N: 'static + ScalarT,
    D: DimName,
    DefaultAllocator: Allocator<N, D>,
{
    type Field = N;
}


examples/circle_example.rs
//! A simple circle example demonstrating how to obtain an 
//! interpolated value at an arbitrary location on a curve.

use nalgebra::Vector2;
use capstan::{knotvec::KnotVec};
type Curve = capstan::curve::Curve<f32, Vector2<f32>>;

fn main() {

    let r = f32::sqrt(2.0) / 2.0;
    let degree = 2;
    let control_points = vec![
        Vector2::new(1.0, 0.0),
        Vector2::new(1.0, 1.0),
        Vector2::new(0.0, 1.0),
        Vector2::new(-1.0, 1.0),
        Vector2::new(-1.0, 0.0),
        Vector2::new(-1.0, -1.0),
        Vector2::new(0.0, -1.0),
        Vector2::new(1.0, -1.0),
        Vector2::new(1.0, 0.0),
    ];
    let weights = vec![1.0, r, 1.0, r, 1.0, r, 1.0, r, 1.0];
    let knots = KnotVec::new(vec![
        0.0, 0.0, 0.0, 0.25, 0.25, 0.5, 0.5, 0.75, 0.75, 1.0, 1.0, 1.0,
    ])
    .unwrap();

    let circle = Curve::new(degree, control_points, weights, knots).unwrap();

    let u = 0.5_f32;
    println!("interpolation paramater value {}:{}", u, circle.de_boor(u))
}

src/main.rs
extern crate svg;

use nalgebra::Vector2;
use svg::node::element::path;
use svg::node::element::Circle;
use svg::node::element::Group;
use svg::node::element::Path;
use svg::node::Node;
use svg::Document;

use capstan::knotvec::KnotVec;
type Curve = capstan::curve::Curve<f32, Vector2<f32>>;

fn main() {
    println!("Plotting some examples");

    circle_example(&String::from("circle.svg"));
    cubic_bezier_example(&String::from("cubic-bezier.svg"));
    reed_leaf_example(&String::from("reed-leaf.svg"));
}

fn reed_leaf_example(filename: &str) {
    let n_div = 256;
    let nurbs_and_cp_group =
        curve_and_control_polygon(&reed_leaf(), n_div).set("transform", "translate(-60,0)");
    let nurbs_group = curve_path(&reed_leaf(), n_div).set("transform", "translate(90, 0)");

    let document = Document::new()
        .set("width", 300)
        .set("height", 300)
        .add(nurbs_and_cp_group)
        .add(nurbs_group);

    svg::save(filename, &document).unwrap();
}

fn circle_example(filename: &str) {
    let radius = 130.0;

    let mut nurbs_circle = unit_circle();
    nurbs_circle.uniform_scale(radius);
    let nurbs_group =
        curve_and_control_polygon(&nurbs_circle, 256).set("transform", "translate(150, 150)");

    let circle = style_regular(Circle::new().set("cx", 0).set("cy", 0).set("r", radius))
        .set("transform", "translate(450, 150)");

    let document = Document::new()
        .set("width", 600)
        .set("height", 300)
        .add(nurbs_group)
        .add(circle);

    svg::save(filename, &document).unwrap();
}

fn cubic_bezier_example(filename: &str) {
    let bezier = style_regular(
        Path::new().set(
            "d",
            path::Data::new()
                .move_to((80, 20))
                .cubic_curve_to((280, 280, 20, 280, 220, 20)),
        ),
    )
    .set("transform", "translate(300, 0)");

    let nurb_bezier = curve_and_control_polygon(
        &Curve::new(
            3,
            vec![
                Vector2::new(80.0, 20.0),
                Vector2::new(280.0, 280.0),
                Vector2::new(20.0, 280.0),
                Vector2::new(220.0, 20.0),
            ],
            vec![1.0, 1.0, 1.0, 1.0],
            KnotVec::new(vec![0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0]).unwrap(),
        )
        .unwrap(),
        256,
    );

    let document = Document::new()
        .set("width", 600)
        .set("height", 300)
        .add(nurb_bezier)
        .add(bezier);

    svg::save(filename, &document).unwrap();
}

fn style_regular<T>(node: T) -> Group
where
    T: Node,
{
    Group::new()
        .add(node)
        .set("fill", "none")
        .set("stroke", "blue")
        .set("stroke-width", "2px")
        .set("vector-effect", "non-scaling-stroke")
}

fn curve_and_control_polygon(curve: &Curve, n_divisions: usize) -> Group {
    Group::new()
        .add(curve_path(curve, n_divisions))
        .add(curve_polygon(curve))
}

fn curve_path(curve: &Curve, n_divisions: usize) -> Path {
    Path::new()
        .set("d", curve_path_data(curve, n_divisions))
        .set("fill", "none")
        .set("stroke", "#711081")
        .set("stroke-width", "2px")
        .set("vector-effect", "non-scaling-stroke")
}

fn curve_path_data(curve: &Curve, n_divisions: usize) -> path::Data {
    let min_u = curve.knots().min_u();
    let max_u = curve.knots().max_u();
    let u_range = max_u - min_u;
    let range_denom = n_divisions as f32;

    let mut commands = Vec::with_capacity(n_divisions + 1);
    commands.push(path::Command::Move(
        path::Position::Absolute,
        path::Parameters::from(eval_curve_2d(&curve, min_u)),
    ));
    for i in 1..(n_divisions + 1) {
        let u = min_u + (i as f32) * u_range / range_denom;
        commands.push(path::Command::Line(
            path::Position::Absolute,
            path::Parameters::from(eval_curve_2d(&curve, u)),
        ))
    }

    path::Data::from(commands)
}

fn curve_polygon(curve: &Curve) -> Group {
    // control points
    let cps = curve.control_points();

    // a group for the control points
    let mut control_points_group = Group::new();
    for cp in cps {
        let cp_circle = control_point(cp.x, cp.y, 3.5);
        control_points_group = control_points_group.add(cp_circle);
    }

    // the control polygon lines
    let mut commands = Vec::with_capacity(cps.len());
    commands.push(path::Command::Move(
        path::Position::Absolute,
        path::Parameters::from((cps[0].x, cps[0].y)),
    ));
    for cp in cps.iter().skip(1) {
        commands.push(path::Command::Line(
            path::Position::Absolute,
            path::Parameters::from((cp.x, cp.y)),
        ));
    }
    let path_data = path::Data::from(commands);
    let path = Path::new()
        .set("d", path_data)
        .set("fill", "none")
        .set("stroke", "#101010")
        .set("stroke-width", "1px")
        .set("stroke-dasharray", "4 3")
        .set("vector-effect", "non-scaling-stroke");

    Group::new().add(path).add(control_points_group)
}

fn control_point(x: f32, y: f32, radius: f32) -> Circle {
    Circle::new()
        .set("cx", x)
        .set("cy", y)
        .set("r", radius)
        .set("fill", "#AAAAAA")
        .set("stroke", "#000000")
        .set("stroke-width", "1px")
        .set("vector-effect", "non-scaling-stroke")
}

fn eval_curve_2d(curve: &Curve, u: f32) -> (f32, f32) {
    let pt_3d = curve.de_boor(u);
    (pt_3d.x, pt_3d.y)
}

fn unit_circle() -> Curve {
    let r = f32::sqrt(2.0) / 2.0;
    let degree = 2;
    let control_points = vec![
        Vector2::new(1.0, 0.0),
        Vector2::new(1.0, 1.0),
        Vector2::new(0.0, 1.0),
        Vector2::new(-1.0, 1.0),
        Vector2::new(-1.0, 0.0),
        Vector2::new(-1.0, -1.0),
        Vector2::new(0.0, -1.0),
        Vector2::new(1.0, -1.0),
        Vector2::new(1.0, 0.0),
    ];
    let weights = vec![1.0, r, 1.0, r, 1.0, r, 1.0, r, 1.0];
    let knots = KnotVec::new(vec![
        0.0, 0.0, 0.0, 0.25, 0.25, 0.5, 0.5, 0.75, 0.75, 1.0, 1.0, 1.0,
    ])
    .unwrap();
    Curve::new(degree, control_points, weights, knots).unwrap()
}

fn reed_leaf() -> Curve {
    let degree = 3;
    let control_points = vec![
        Vector2::new(152.0, 18.0),
        Vector2::new(140.0, 24.0),
        Vector2::new(130.0, 29.0),
        Vector2::new(121.0, 41.0),
        Vector2::new(105.0, 65.0),
        Vector2::new(105.0, 96.0),
        Vector2::new(107.0, 282.0),
        Vector2::new(107.0, 282.0),
        Vector2::new(125.0, 277.0),
        Vector2::new(125.0, 277.0),
        Vector2::new(125.0, 267.0),
        Vector2::new(123.0, 235.0),
        Vector2::new(129.0, 230.0),
        Vector2::new(140.0, 221.0),
        Vector2::new(158.0, 209.0),
        Vector2::new(173.0, 201.0),
        Vector2::new(173.0, 201.0),
        Vector2::new(152.0, 18.0),
        Vector2::new(152.0, 18.0),
    ];
    let weights = vec![
        1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
        1.0,
    ];
    let knots = KnotVec::new(vec![
        0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 5.0, 5.0,
        5.0, 6.0, 6.0, 6.0, 6.0,
    ])
    .unwrap();
    Curve::new(degree, control_points, weights, knots).unwrap()
}
